ouroboros-network/src/Ouroboros/Network/Block.hs

```diff
@@ -108,27 +116,37 @@ castHash (BlockHash b) = BlockHash b
   Point on a chain
 -------------------------------------------------------------------------------}
 
--- | A point on the chain is identified by its 'Slot' and 'HeaderHash'.
---
--- The 'Slot' tells us where to look and the 'HeaderHash' either simply serves
--- as a check, or in some contexts it disambiguates blocks from different forks
--- that were in the same slot.
---
-data Point block = Point {
-       pointSlot :: SlotNo,
-       pointHash :: ChainHash block
-     }
-   deriving (Eq, Ord, Show)
+-- | A block is identified by a slot and a hash.
+data TBlockPoint slot hash = TBlockPoint
+  { pointSlot :: slot
+  , pointHash :: hash
+  }
+  deriving (Eq, Ord, Show)
+
+type BlockPoint block = TBlockPoint SlotNo (HeaderHash block)
+
+-- | `BlockPoint` with an origin point.
+data TPoint t where
+  Origin :: TPoint t
+  Point  :: t -> TPoint t
+  deriving (Eq, Ord, Show)
+
+instance Functor TPoint where
+  fmap _ Origin    = Origin
+  fmap f (Point t) = Point (f t)
+
+fromTPoint :: r -> (t -> r) -> TPoint t -> r
+fromTPoint r _ Origin    = r
+fromTPoint _ f (Point t) = f t
 
-castPoint :: (HeaderHash a ~ HeaderHash b) => Point a -> Point b
-castPoint (Point a b) = Point a (castHash b)
+type Point block = TPoint (BlockPoint block)
```

Comment:
But why do we need the two variants of `Point`? Why not simply:
```
data Point block = GenesisPoint | Point (BlockPoint block)
```
Is it really that much of a benefit to it being a functor? Can we handle those (presumably rather rare) cases by conversion to/from `Maybe (BlockPoint block)`?

Fewer names! Fewer concepts :-)

Reply:
I will make a change to get rid of the `T`-prefix. It's there because I wanted less of a fallout from a name change.

> Is it really that much of a benefit to it being a functor?

Yes. It's used in `Ouroboros.Network.Chain.{headSlot, headHash}` rather effectively. Again in `Ouroboros.Network.AnchoredFragment.{headSlot, lastSlot}`.

> Fewer names! Fewer concepts :-)

In my experience this is not always better. It can in fact make things worse. Why have nominal types at all?